File: lab-experiments/cell-sedimentation/src/ts/calcs.ts
import type { Profile } from "../types/globals"
import { rk45 } from "./rk45"

const rho_f = 0.868 // Fluid density (g/cc)
const g = 981 // Gravitational acceleration (cm/s^2)
const mu = 2.15 // Fluid viscosity (mPa.s)

const rho_r = 1.08 // Red cell density (g/cc)
const d_r = 275e-4 // Red cell diameter (cm)
const rho_w = 1.00 // White cell density (g/cc)
const d_w = 550e-4 // White cell diameter (cm)

export const TUBE_LENGTH = 305; // mm
export const CONC_ARRAY_SIZE = 500; // points in finite element mesh

/**
 * Convert volume fraction to number concentration (#/[l]^3) for particles of diameter dp units [l].
 * @param phi volume fraction of particle
 * @param dp diameter of particle
 */
function concentration(phi: number, dp: number) {
    return 6 * phi / (Math.PI * dp**3);
}
// Wrappers for more concise syntax
export const conc_r = (phi: number) => concentration(phi, d_r);
export const conc_w = (phi: number) => concentration(phi, d_w);

const cr_max = conc_r(1);
const cw_max = conc_w(1);

/**
 * Convert concentration (#/cc) to volume fraction for particles of diameter dp (cm).
 * @param conc 
 * @param dp 
 * @returns 
 */
function volume_fraction(conc: number, dp: number) {
    return conc * Math.PI * dp**3 / 6;
}
// Wrappers for more concise syntax
export const phi_r = (phi: number) => volume_fraction(phi, d_r);
export const phi_w = (phi: number) => volume_fraction(phi, d_w);

/**
 * Calculate the settling velocity of an independant particle using Stokes' law.
 * @param d_p 
 * @param rho_p 
 * @returns 
 */
function settling_coeff_g(d_p: number, rho_p: number) {
    const vs = (rho_p - rho_f) * g * d_p**2 / (18 * mu) // m/s
    return vs * 1000 // Convert to mm/s
}

// Average void envelope thickness
function d_Ep(r_conc: number, w_conc: number) {
    if (r_conc + w_conc === 0) return 0;

    const d_avg = (r_conc * d_r + w_conc * d_w)  / (r_conc + w_conc); // Average particle diameter
    const ep = 1.0 - (r_conc + w_conc) * (1 - .4) // Typical packed bed void fraction for spheres
    if (ep >= 1) return 0;

    const pow = (1 - ep) ** (-1/3)
    return d_avg * (pow - 1) // Void envelope thickness
}

// Calculate settling velocities
const Sr_g = settling_coeff_g(d_r, rho_r)
const Sw_g = settling_coeff_g(d_w, rho_w)

// Calculate the effective settling velocity (mm/s) of a mixture of red and white cells.
export function particle_velocities(conc_r: number, conc_w: number) {
    // Convert to volume fractions
    const ph_r = volume_fraction(conc_r, d_r)
    const ph_w = volume_fraction(conc_w, d_w)
    
    const conc_f = 1 - ph_r - ph_w // volume fraction of fluid
    const rho_susp = ph_r * rho_r + ph_w * rho_w + conc_f * rho_f // Suspension density (g/cc)
    const d_ep = d_Ep(ph_r, ph_w) // Average void envelope thickness
    const ep_r = (d_ep > 0) ? 1 - (1 + d_ep / d_r) ** -3 : 1; // Red cell porosity
    const ep_w = (d_ep > 0) ? 1 - (1 + d_ep / d_w) ** -3 : 1; // White cell porosity
    const vrs = Sr_g * (rho_r - rho_susp) / (rho_r - rho_f) * ep_r**4.6 // Red cells hindered settling
    const vws = Sw_g * (rho_w - rho_susp) / (rho_w - rho_f) * ep_w**4.6 // White cells hindered settling
    const vfc = -(ph_r * vrs + ph_w * vws) // Fluid counterflow
    const vrc = vrs + vfc // Red cell velocity
    const vwc = vws + vfc // White cell velocity
    return { red: vrc, white:  vwc } // mm / s
}

/**
 * Calculate spatial derivative of 1D y mesh with x differential dx
 * @param y array of scalar values
 * @param dx spacing between y values
 * @returns dydx array
 */
export function grad(y: number[], dx: number) {
    const len = y.length;
    if (len < 2) return [0];

    // Generate array and set edges
    const dydx = new Array(len);
    dydx[0] = (y[1] - y[0]) / dx;
    dydx[len-1] = (y[len-1] - y[len-2]) / dx;

    // Calc middle derivs with 2nd order accuracy
    for (let i=1; i<len-1; i++) {
        const left = y[i-1];
        const right= y[i+1];
        dydx[i] = (right - left) / 2 / dx;
    }
    
    return dydx;
}

export function rhs_adv_diff(y: number[], dz: number) {
    // Unpack y
    const cr = y.slice(0,CONC_ARRAY_SIZE); 
    const cw = y.slice(CONC_ARRAY_SIZE);

    // BCs
    cr[0] = 0; cw[0] = 0;
    cr[CONC_ARRAY_SIZE - 1] = cr[CONC_ARRAY_SIZE - 2]; 
    cw[CONC_ARRAY_SIZE - 1] = cw[CONC_ARRAY_SIZE - 2];

    // For advection term
    const crvr = new Array(CONC_ARRAY_SIZE);
    const cwvw = new Array(CONC_ARRAY_SIZE);
    crvr[0] = 0;
    cwvw[0] = 0;
    for (let i=1;i<CONC_ARRAY_SIZE;i++) {
        // Determine velocity
        const { red, white } = particle_velocities(cr[i], cw[i]);
        if (red !== red) throw new Error(`bad red at concentration ${cr[i]} ${cw[i]}`)
        crvr[i] = cr[i] * red;
        cwvw[i] = cw[i] * white;
        if (crvr[i] !== crvr[i]) throw new Error(`crvr err: ${crvr}`)
    }
    crvr[CONC_ARRAY_SIZE-1] = 0;
    cwvw[CONC_ARRAY_SIZE-1] = 0;

    // Advection term
    const adv_r = grad(crvr, dz);
    const adv_w = grad(cwvw, dz);
    for (let i=0;i<CONC_ARRAY_SIZE;i++) {
        if (adv_r !== adv_r) throw new Error(`adv_r error: ${adv_r}`)
    }

    // For diffusion term
    const inner_dif_r = new Array(CONC_ARRAY_SIZE);
    const inner_dif_w = new Array(CONC_ARRAY_SIZE);
    inner_dif_r[0] = 5 * (cr[1] - cr[0]) / dz * (1 - cr[0] / cr_max)**10;
    inner_dif_w[0] = 5 * (cw[1] - cw[0]) / dz * (1 - cw[0] / cw_max)**10;
    for (let i=1;i<CONC_ARRAY_SIZE-1;i++) {
        const dcri = (cr[i+1] - cr[i-1]) / 2 / dz;
        const dcwi = (cw[i+1] - cw[i-1]) / 2 / dz;
        inner_dif_r[i] = 5 * dcri * (1 - cr[i] / cr_max)**10;
        inner_dif_w[i] = 5 * dcwi * (1 - cw[i] / cw_max)**10;
        if (inner_dif_r[i] !== inner_dif_r[i]) throw new Error(`inner_dif_r err: ${inner_dif_r}`)
    }
    inner_dif_r[CONC_ARRAY_SIZE-1] = 5 * (cr[CONC_ARRAY_SIZE-1] - cr[CONC_ARRAY_SIZE-2]) / dz * (1 - cr[CONC_ARRAY_SIZE-1] / cr_max)**10;
    inner_dif_w[CONC_ARRAY_SIZE-1] = 5 * (cw[CONC_ARRAY_SIZE-1] - cw[CONC_ARRAY_SIZE-2]) / dz * (1 - cw[CONC_ARRAY_SIZE-1] / cw_max)**10;

    const dif_r = grad(inner_dif_r, dz);
    const dif_w = grad(inner_dif_w, dz);
    dif_r[0] = dif_r[CONC_ARRAY_SIZE-1] = 0;
    dif_w[0] = dif_w[CONC_ARRAY_SIZE-1] = 0;

    // Results 
    const drdt = new Array(CONC_ARRAY_SIZE);
    const dwdt = new Array(CONC_ARRAY_SIZE);
    drdt[0] = dwdt[0] = 0;
    for (let i=1; i<CONC_ARRAY_SIZE-1;i++) {
        drdt[i] = dif_r[i] - adv_r[i];
        dwdt[i] = dif_w[i] - adv_w[i];
    }
    drdt[CONC_ARRAY_SIZE-1] = -adv_r[CONC_ARRAY_SIZE-1];
    dwdt[CONC_ARRAY_SIZE-1] = -adv_w[CONC_ARRAY_SIZE-1];

    return drdt.concat(dwdt);
}

/**
 * Convolve a 1D numpy array y with a moving average filter of given size.
 * Pads the edges with the first and last values in y to maintain size.
 * @param y array to smooth
 * @param window_size size of kernel
 * @returns smoothed array of same length as y
 */
export function movingAverageConvolve(y: number[], windowSize: number): number[] {
  if (windowSize < 1) throw new Error("windowSize must be at least 1");
  if (windowSize === 1) return y;

  const padWidth = Math.floor(windowSize / 2);
  const n = y.length;

  // Pad edges with the first and last values
  const padded = [
    ...Array(padWidth).fill(y[0]),
    ...y,
    ...Array(padWidth).fill(y[n - 1])
  ];

  // Precompute kernel weight
  const norm = 1 / windowSize;

  const smoothed: number[] = new Array(n);

  // Perform convolution
  for (let i = 0; i < n; i++) {
    let sum = 0;
    for (let j = 0; j < windowSize; j++) {
      sum += padded[i + j];
    }
    smoothed[i] = sum * norm;
  }

  return smoothed;
}

/**
 * Linearly interpolate points in x to find corresponding points in y using xp and fp
 * @param x x coordinates of resulting points
 * @param xp x coordinates to interpolate between
 * @param fp values corresponding to original xp array
 * @returns f values corresponding to each x
 */
export function interp(x: number | number[], xp: number[], fp: number[]) {
    if (!Array.isArray(x)) x = [x];
    const f = x.map(xi => {
        let nx = xp.length;
        // Make sure we stay in the bounds
        if (xi < xp[0]) {
            console.warn(`interp point ${xi} out of range for x array with min ${xp[0]}`);
            return fp[0];
        }

        // Find the lo and hi indices
        let lo=0, hi=1;
        for (let j=0;j<nx;j++) {
            if (xp[j] <= xi) lo = j;
            else break;
        }

        // Interpolate results
        if (lo === nx-1) {
            console.warn(`interp point ${xi} out of range for x array with min ${xp[nx-1]}`);
            return fp[nx-1];
        }
        hi = lo + 1;
        const fi = fp[lo] + (fp[hi] - fp[lo]) * (xi - xp[lo]) / (xp[hi] - xp[lo]);
        return fi;
    });

    return f;
}

/**
 * 
 * @param y 
 * @param z 
 * @param lo 
 * @param hi 
 * @returns 
 */
export function resize(y: number[], z: number[], lo: number) {
    const cr = y.slice(0, CONC_ARRAY_SIZE);
    const cw = y.slice(CONC_ARRAY_SIZE);

    if (cr.length !== z.length || cw.length !== z.length) throw new Error(`y and z must have same length\ncr:${cr.length}\ncw:${cw.length}\nz:${z.length}`);
    const nz = cr.length;
    const z_hi = z[nz - 1];
    if (z_hi !== 305) throw new Error("double check z my guy");

    // Determine where to start the new coordinate axis
    let min_idx: number | undefined = undefined;
    let sum = 0;
    for (let i=0;i<cr.length;i++) {
        const ph = cr[i] + cw[i];
        sum += ph;
        if (sum < lo) min_idx = i;
        else if (sum >= 2*lo) break;
    }
    if (min_idx === undefined) return { y, z };

    // Determine the new range and construct the new z array
    const lz = z_hi - z[min_idx];
    const z_new = Array.from({ length: nz }, (_,i) => i / (nz-1) * lz + z[min_idx]);

    // Interpolate to get the new f(z) points in the array.
    const cr_new = interp(z_new, z, cr);
    const cw_new = interp(z_new, z, cw);
    
    return { y: cr_new.concat(cw_new), z: z_new };
}


const nz = CONC_ARRAY_SIZE;
const smooth_fsize = 13;
const dt = 20;
export class ProfileSolver {
    private current: Profile;
    private top = 0;

    constructor(xr0: number, xw0: number) {
        // Prepare the initial profile
        const cr0: number[] = new Array(nz).fill(conc_r(xr0));
        const cw0: number[] = new Array(nz).fill(conc_w(xw0));
        cr0[0] = cw0[0] = 0; // Free surface
        this.current = cr0.concat(cw0);

        
    }

    // Spatial helpers
    private lz = () => { return TUBE_LENGTH - this.top; }
    private dz = () => { return this.lz() / (nz - 1); }
    private create_z_arr = () => {
        const dz = this.dz();
        return Array.from({ length: nz }, (_,i) => i * dz);
    }

    // Rhs wrapper
    private rhs = (_: number, y: number[]) => {
        return rhs_adv_diff(y, this.dz());
    };

    /**
     * Calculate one 20 second step for the profile
     * @returns Results at time t + dt
     */
    public calculate_step = () => {
        const sol = rk45(this.rhs, this.current, 0, dt);

        // Extract solution
        let y_cur: Profile = sol.y.at(-1)!;
        let cr_cur = y_cur.slice(0,nz);
        let cw_cur = y_cur.slice(nz);

        // Smooth
        cr_cur = movingAverageConvolve(cr_cur, smooth_fsize);
        cw_cur = movingAverageConvolve(cw_cur, smooth_fsize);
        y_cur = cr_cur.concat(cw_cur);

        // Resize
        const resized = resize(y_cur, this.create_z_arr(), .05);
        this.current = resized.y;
        this.top = resized.z[0];

        return this.current.concat([this.top]);
    }
}

File: lab-experiments/cell-sedimentation/src/wasm/calcs.cpp
#include "calcs.h"
#include <stdexcept>

inline double d_Ep(double r_conc, double w_conc) {
	if (r_conc + w_conc == 0.0) return 0.0;
	double d_avg = (r_conc * DP_R + w_conc * DP_W) / (r_conc + w_conc);
	double ep = 1.0 - (r_conc + w_conc) * (1.0 - 0.4);
	if (ep >= 1.0) return 0.0;
	double powv = std::pow(1.0 - ep, -1.0/3.0);
	return d_avg * (powv - 1.0);
}

static const double INV_RHO_DIFF_R = 1.0 / (RHO_R - RHO_F);
static const double INV_RHO_DIFF_W = 1.0 / (RHO_W - RHO_F);

// build once at program init
#define EP_N 1024
static double ep_pow_table[EP_N];
static double ep_min = 1e-8;

void init_ep_table() {
    for (int i=0;i<EP_N;i++){
        double e = ep_min + (1.0 - ep_min) * i / (EP_N-1);
        ep_pow_table[i] = std::pow(e, 4.6);
    }
}

inline double ep_pow_lookup(double ep) {
    if (ep <= ep_min) return ep_pow_table[0];
    if (ep >= 1.0) return 1.0;
    double t = (ep - ep_min) / (1.0 - ep_min) * (EP_N - 1);
    int i = (int) t;
    double frac = t - i;
    return (1.0 - frac) * ep_pow_table[i] + frac * ep_pow_table[i+1];
}

inline Velocities particle_velocities(double conc_r, double conc_w) {
    double ph_r = phi_r(conc_r);
    double ph_w = phi_w(conc_w);
    if (ph_r + ph_w >= 1.0) return {0.0, 0.0};

    double conc_f = 1.0 - ph_r - ph_w;
    double rho_susp = ph_r * RHO_R + ph_w * RHO_W + conc_f * RHO_F;

    double d_ep = d_Ep(ph_r, ph_w);
    double ep_r = (d_ep > 0.0) ? 1.0 - std::pow(1.0 + d_ep / DP_R, -3.0) : 1.0;
    double ep_w = (d_ep > 0.0) ? 1.0 - std::pow(1.0 + d_ep / DP_W, -3.0) : 1.0;

    // clamp to avoid log(0) in approximations
    ep_r = std::fmax(ep_r, 1e-12);
    ep_w = std::fmax(ep_w, 1e-12);

    double vrs = SR_G * (RHO_R - rho_susp) * INV_RHO_DIFF_R * ep_pow_lookup(ep_r);
    double vws = SW_G * (RHO_W - rho_susp) * INV_RHO_DIFF_W * ep_pow_lookup(ep_w);

    double vfc = -(ph_r * vrs + ph_w * vws);
    return { vrs + vfc, vws + vfc };
}

double grad(RunningWindow y, double dx) {
	double dydx = (y.next - y.last) / 2.0 / dx;
	return dydx;
}

void shift(RunningWindow& win, double newValue) {
    win.last = win.curr;
    win.curr = win.next;
    win.next = newValue;
}

void movingAverageConvolve(double* y, int n, int windowSize) {
    if (windowSize <= 1) return; // no change

    double* result = new double[n];
    int halfWindow = windowSize / 2;

    for (int i = 0; i < n; ++i) {
        double sum = 0.0;
        for (int j = -halfWindow; j <= halfWindow; ++j) {
            int idx = i + j;
            if (idx >= 0 && idx < n) {
                sum += y[idx];
            }
            else if (idx < 0) {
                sum += y[0]; // replicate edge
            }
            else {
                sum += y[n-1]; // replicate edge
            }
        }
        result[i] = sum / windowSize;
    }

    // Copy back to original array
    for (int i = 0; i < n; ++i) {
        y[i] = result[i];
    }

    delete[] result;
}
    
double interp(double xp, double x_lo, double x_hi, double* y, int n) {
    double range = x_hi - x_lo;
    xp = (xp - x_lo) / range * (n - 1);
    int i_lo = (int)xp;
    int i_hi = i_lo + 1;
    if (i_lo < 0) return y[0];
    if (i_hi >= n) return y[n-1];
    double denom = (double)(i_hi - i_lo);
    if (denom == 0.0) return y[i_lo];

    // Linear interpolation
    double fi = y[i_lo] + (y[i_hi] - y[i_lo]) * (xp - (double)i_lo) / denom;
    return fi;
}

double calc_dz(double lz) {
    return lz / (CONC_ARRAY_SIZE - 1);
}

// void rhs_adv_diff(const ConcPair y[], ConcPair out[], double dz) {
//     // Pack Left BC in. First iteration will shift out "last" values.
//     RunningWindow crvr = { 0.0, 0.0, 0.0 };
//     RunningWindow cwvw = { 0.0, 0.0, 0.0 };
//     RunningWindow inner_dif_r = { 0.0, 0.0, 5 * y[1].r / dz };
//     RunningWindow inner_dif_w = { 0.0, 0.0, 5 * y[1].w / dz };

//     for (size_t i = 0; i < CONC_ARRAY_SIZE-2; ++i) {
//         // Unpack y
//         double cr = y[i].r;
//         double cw = y[i].w;
//         double cr_next = y[i + 1].r;
//         double cw_next = y[i + 1].w;

//         // Calculate crvr and cwvw
//         Velocities vels = particle_velocities(cr_next, cw_next);
//         shift(crvr, cr_next * vels.red);
//         shift(cwvw, cw_next * vels.white);

//         // Use to calculate advection terms
//         double adv_r = grad(crvr, dz);
//         double adv_w = grad(cwvw, dz);

//         // Calculate inner diffusion terms
//         RunningWindow cr_next_3 = { cr, cr_next, y[i + 2].r };
//         RunningWindow cw_next_3 = { cw, cw_next, y[i + 2].w };
//         double dr_next = grad(cr_next_3, dz);
//         double dw_next = grad(cw_next_3, dz);
//         shift(inner_dif_r, 5 * dr_next * std::pow((1.0 - cr_next / CR_MAX), 10));
//         shift(inner_dif_w, 5 * dw_next * std::pow((1.0 - cw_next / CW_MAX), 10));
//         double dif_r = grad(inner_dif_r, dz);
//         double dif_w = grad(inner_dif_w, dz);

//         // Combine to get rhs
//         out[i].r = -adv_r + dif_r;
//         out[i].w = -adv_w + dif_w;
//     }
//     out[0].r = 0.0;
//     out[0].w = 0.0;
//     // Compute second to last point (i = CONC_ARRAY_SIZE - 2)
//     shift(crvr, 0.0);
//     shift(cwvw, 0.0);
//     shift(inner_dif_r, 0);
//     shift(inner_dif_w, 0);
//     double adv_r = grad(crvr, dz);
//     double adv_w = grad(cwvw, dz);
//     double dif_r = (inner_dif_r.next - inner_dif_r.last) / 2 / dz;
//     double dif_w = (inner_dif_w.next - inner_dif_w.last) / 2 / dz;
//     out[CONC_ARRAY_SIZE - 2].r = -adv_r + dif_r;
//     out[CONC_ARRAY_SIZE - 2].w = -adv_w + dif_w;

//     // Last element
//     out[CONC_ARRAY_SIZE - 1].r = out[CONC_ARRAY_SIZE - 2].r;
//     out[CONC_ARRAY_SIZE - 1].w = out[CONC_ARRAY_SIZE - 2].w;
// }

#define PAD_SIZE 2

void rhs_adv_diff(const ConcPair y[], ConcPair out[], double dz) {
    // // Pack Left BC in. First iteration will shift out "last" values.
    ConcPair y_padded[CONC_ARRAY_SIZE + 2 * PAD_SIZE];
    ConcPair cv_arr[CONC_ARRAY_SIZE + 2];
    ConcPair inner_dif_arr[CONC_ARRAY_SIZE + 2];

    // Pad y with BCs
    y_padded[0] = {0.0, 0.0};
    y_padded[1] = {0.0, 0.0};
    for (int i = 0; i < CONC_ARRAY_SIZE; ++i) {
        y_padded[i + PAD_SIZE] = y[i];
    }
    y_padded[CONC_ARRAY_SIZE + PAD_SIZE] = {0.0, 0.0};
    y_padded[CONC_ARRAY_SIZE + PAD_SIZE + 1] = {0.0, 0.0};

    // First stage: Calculate all cv values and inner diffusion terms.
    // #pragma omp parallel for
    for (int i = 0; i < CONC_ARRAY_SIZE; ++i) {
        int idx = i + PAD_SIZE; // account for padding

        double cr = y_padded[idx].r;
        double cw = y_padded[idx].w;

        // Calculate crvr and cwvw
        Velocities vels = particle_velocities(cr, cw);
        cv_arr[i + 1] = { cr * vels.red, cw * vels.white };

        // Calculate inner diffusion terms
        double dr = 5.0 * (y_padded[idx + 1].r - y_padded[idx - 1].r) / (2 * dz) * std::pow(1.0 - cr / CR_MAX, 10);
        double dw = 5.0 * (y_padded[idx + 1].w - y_padded[idx - 1].w) / (2 * dz) * std::pow(1.0 - cr / CR_MAX, 10);
        inner_dif_arr[i + 1] = { dr, dw };
    }

    // Pad cv_arr and inner_dif_arr for BCs
    cv_arr[0] = {0.0, 0.0};
    cv_arr[1] = {0.0, 0.0};
    cv_arr[CONC_ARRAY_SIZE] = {0.0, 0.0};
    cv_arr[CONC_ARRAY_SIZE + 1] = {0.0, 0.0};

    inner_dif_arr[0] = {0.0, 0.0};
    // inner_dif_arr[1] = { 5 * y[0].r / dz, 5 * y[0].w / dz };
    inner_dif_arr[CONC_ARRAY_SIZE] = { 0, 0 };
    inner_dif_arr[CONC_ARRAY_SIZE + 1] = {0.0, 0.0};

    // Second stage: Calculate advection and diffusion gradients and combine.
    for (int i = 0; i < CONC_ARRAY_SIZE; ++i) {
        // int y_idx = i + PAD_SIZE; // account for padding
        int cv_idx = i + 1; // account for cv_arr padding

        // Advection gradients
        double adv_r = (cv_arr[cv_idx + 1].r - cv_arr[cv_idx - 1].r) / (2 * dz);
        double adv_w = (cv_arr[cv_idx + 1].w - cv_arr[cv_idx - 1].w) / (2 * dz);

        // Diffusion gradients
        double dif_r = (inner_dif_arr[cv_idx + 1].r - inner_dif_arr[cv_idx - 1].r) / (2 * dz);
        double dif_w = (inner_dif_arr[cv_idx + 1].w - inner_dif_arr[cv_idx - 1].w) / (2 * dz);

        // Combine to get rhs
        out[i].r = -adv_r + dif_r;
        out[i].w = -adv_w + dif_w;
    }

    out[0] = { 0, 0 };
    out[CONC_ARRAY_SIZE - 1] = out[CONC_ARRAY_SIZE - 2];
}

double resize(double* y, double top, double lo) {
	double cr[CONC_ARRAY_SIZE];
	double cw[CONC_ARRAY_SIZE];
    for (size_t i = 0; i < CONC_ARRAY_SIZE; ++i) {
        cr[i] = y[i];
        cw[i] = y[i + CONC_ARRAY_SIZE];
    }

    double lz = 305 - top;
	int min_idx = -1;
	double sum = 0.0;
    double min_z = top;
	for (size_t i = 0; i < CONC_ARRAY_SIZE; ++i) {
        double zi = (double)i / (double)(CONC_ARRAY_SIZE - 1) * lz + top;
		double ph = phi_r(cr[i]) + phi_w(cw[i]);
		sum += ph;
		if (sum < lo) {
            min_idx = (int)i;
            min_z = zi; // store the actual z value at min_idx. Eventually return as new top
        }
		else if (sum >= 2.0 * lo) break;
	}
	if (min_idx == -1) return min_z; // no resizing needed

    double nlz = 305 - min_z;
    // Resize concentration profiles
	for (size_t i = 0; i < CONC_ARRAY_SIZE; ++i) {
        double z_new = (double)i / (double)(CONC_ARRAY_SIZE - 1) * nlz + min_z;
        double cr_new = interp(z_new, top, 305.0, cr, CONC_ARRAY_SIZE);
        double cw_new = interp(z_new, top, 305.0, cw, CONC_ARRAY_SIZE);
        y[i] = cr_new;
        y[i + CONC_ARRAY_SIZE] = cw_new;
    }

	return min_z;
}

